### PR TITLE
split outputs, build with rattler-build, flang 19

### PR DESCRIPTION
**Pull request by minrk**
use rattler because this is a lot of outputs and it saves a lot of time for the many env setup/teardown transactions over conda-build

adds suitesparse-mongoose as an additional component and unskips `spex` on windows, now that the required `gmp` is available there.

Further splitting _not_ done, that could be, and I _think_ could be done later without major consequence:

- `suitesparse-devel` to contain headers/cmake/pkg-config files (I think this needs the approach in #99, which conda-smithy can't handle yet)
- mongoose is both a lib and executable so could be split further, but since it's one package and doesn't support component install, I did not make `libsuitesparse-mongoose`, only `suitesparse-mongoose` with bin and lib.

- [x] needs https://github.com/conda-forge/admin-requests/pull/1193 before the new outputs can be published

closes #100
closes #99

### CODE REVIEW

**Comment by minrk** [comment]
lint error is https://github.com/conda-forge/conda-smithy/issues/2165 already fixed by https://github.com/conda-forge/conda-smithy/pull/2166

**Comment by minrk** [comment]
This actually runs all the builds independently in sequence. A faster approach would be to keep the single 'build and install everything' and use `files`. Downside of `files`: we need to keep track of which files go with which package, and know about every added file for each release. The approach in this PR is a lot easier to keep correct, at the expense of some time (still only 5 minutes on linux-64 thanks to the switch to rattler-build).

**Comment by xhochy** [comment]
At this point you could also think of splitting it into `libpkg` and `libpkg-devel`. If you do it at a later point, you can no longer use `libpkg` as the package containing only the runtime library as this would probably break existing packages (too much).

**Comment by minrk** [comment]
@xhochy that's a good point. I did think of that and figured the cost of the dev files wasn't big enough to justify the recipe complexity. It's a little hard with this multi-package layout where multiple builds have to run unless I go with a single build and pure `files` splitting since I don't think I can run a bunch of builds and do files splitting from several different build steps.

One way to do the splitting that I think would be maintainable is to switch to a single build and do files splitting where:

- lib* packages include only libs (a simple glob for each)
- a single `suitesparse-devel` includes _all_ the headers/cmake files, making the file-include patterns tractable (include/, lib/cmake/ etc.)

so that a package depending on e.g. umfpack would have host:

```yaml
- libumfpack # lib, run_exports
- suitesparse-devel # headers, etc., no run_exports
```

Do you think that's worth it?

Some numbers for the [fully installed suitesparse](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fosx-arm64%2Fsuitesparse-7.8.3-h12cb078_1.conda) to help inform the choice, which should be unchanged (plus the addition of suitesparse-mongoose):

- total instal size: 5.4MB, 159 files
- libs: 4.1MB, 51 files (2/3 are symlinks)
- include: 800kB, 19 files
- cmake, pkg-config: 500kB, 89 files

so the devel package would reduce the number of files for a full suitesparse install by 2/3 and the size by only 20%.

The single-build would also have to wait for https://github.com/conda-forge/conda-smithy/issues/2057 to resolve because rattler-build currently has the build cache behind an `--experimental` flag we can't yet pass.

**Comment by xhochy** [comment]
> @xhochy that's a good point. I did think of that and figured the cost of the dev files wasn't big enough to justify the recipe complexity. 

That is something where I don't have an opinion on as I am not a user of this package, but I stumbled on this case of `libpkg`  vs `libpkg-devel` several times in the last days and wanted to raise it again, so that if you would have long-term plans for it, that you don't "fall into the same trap":

**Comment by minrk** [comment]
Since most packages don't split `-devel` on conda-forge, I've mostly been doing it considering the size/quantity of devel files compared to both the library itself, and the size/frequency of a typical env that depends on it. 1MB of headers in an env that's unlikely to be less that 300MB due to other dependencies doesn't seem entirely worth it.

I also don't think splitting further `libumfpack` into `libumfpack-devel` down the line is likely to break anything, as it would only mean the next _build_ depending on libumfpack would need to add `-devel`, but runtime would only see an install-size savings without any breakage. Or is there something I'm missing there?

**Comment by minrk** [comment]
@conda-forge-admin please lint

File: recipe/recipe.yaml
Comment (wolfv) [inline recipe/recipe.yaml:144]: I believe `build-one` would be enough, and rattler-build will add the appropriate `.sh` or `.bat` ending :)

File: recipe/recipe.yaml
Comment (wolfv) [inline recipe/recipe.yaml:152]: nice 👀

Comment (wolfv) [reply recipe/recipe.yaml:152]: I am sure you know, but inline jinja should also work (`${{ "ninja" if win else "make" }}`)

File: recipe/recipe.yaml
Comment (minrk) [reply recipe/recipe.yaml:144]: Awesome! I tried to figure this out in the docs, but couldn't find an example. An example of this in the multi-output docs would be fantastic.

File: recipe/recipe.yaml
Comment (wolfv) [reply recipe/recipe.yaml:153]: For rattler-build it doesn't really matter as we have to parse the `jinja` AST in both cases anyways (the `if: <cond>` is also evaluated with minijinja, and we extract any used variables from there before rendering the recipe).

Comment (minrk) [reply recipe/recipe.yaml:153]: Yeah, I had the model in mind that when conda-smithy might be collecting the list of dependencies, it can concatenate all the lists and ignore the conditionals, assuming every branch will be taken by _some_ build. But that may not have anything to do with what actually happens. And presumably it still needs to evaluate jinja for every build matrix entry anyway for completeness, so it probably doesn't make a difference.

I just get the _feeling_ that it's nice to avoid templates where possible in the new format, even if it doesn't actually matter in practice.

**Comment by minrk** [comment]
This feels like a big migration to do by myself, so I'd like some approvals before merging. But at this point, I'm not sure who from core and/or @conda-forge/suitesparse should weigh in, but any 👍s would be appreciated.

File: recipe/recipe.yaml
Comment (xhochy) [inline recipe/recipe.yaml:14]: What's the reason to duplicate the version into the build string?

File: recipe/recipe.yaml
Comment (xhochy) [inline recipe/recipe.yaml:34]: Personal opinion: I would just use ninja for all platforms as errors are better readable with it.

File: recipe/recipe.yaml
Comment (xhochy) [inline recipe/recipe.yaml:79]: I'm confused? Why is there a run_exports section in here? What does it do? 

Hint: I have not created complex recipes with v1 syntax.

**Review by xhochy — COMMENTED** [review]
Looks good, just some questions for my own understanding.

File: recipe/recipe.yaml
Comment (minrk) [reply recipe/recipe.yaml:14]: suitesparse is a 'suite' of related but independent libraries. The suite has a version, but each library _also_ has its own version. This just makes it easier to see that `libumfpack 6.3.5` was built as part of `libsuitesparse 7.8.3`. Not all libraries will get any version bump when there is a new suitesparse release, so there may be an ss783 build of libumfpack 6.3.5 _and_ an `ss784` build of libumfpack 6.3.5. I don't expect this will be used for dependency restrictions or anything, it is mostly for clarity and to ensure each build is uploaded even if the library version and hash don't change.

File: recipe/recipe.yaml
Comment (minrk) [reply recipe/recipe.yaml:79]: Each library output should have its own `run_exports` because the idea behind this split is that ~everything depending on `suitesparse` now should migrate to depending directly on the subset of libraries they actually use (`libumfpack`, etc.). If the question is why it's under `requirements`, v1 moves `build.run_exports` to `requirements.run_exports`.

**Review by xhochy — APPROVED** [review]
Thank you for clarifying things!

File: recipe/recipe.yaml
Comment (h-vetinari) [inline recipe/recipe.yaml:60]: Could we automate how the package-specific versions get populated? Or at least auto-validate it based on some in-repo information?

I feel it'll be tricky to update multiple versions correctly if everything is done manually (and the bot will not do it either, leading to an even more tempting "just merge a green PR" case where this could happen)

File: recipe/build-one.bat
Comment (h-vetinari) [inline recipe/build-one.bat:10]: I'd kinda prefer switching to newer flang already (copy [this](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/migrations/flang19.yaml) under `.ci_support/migrations` and rerender), because then you can just do the same logic as in the unix script (e.g. test if `FC` is not empty).

File: recipe/build-one.sh
Comment (h-vetinari) [inline recipe/build-one.sh:42]: OK, just saw that the subpackage versions are validated here (somewhat implicitly) using the SOVERSIONS. The problem is that this won't trigger if upstream bumps the version but doesn't change the SOVERSION (not sure how realistic this is; it may be good enough as-is).

**Review by h-vetinari — COMMENTED** [review]
just some minor questions

File: recipe/recipe.yaml
Comment (minrk) [reply recipe/recipe.yaml:60]: > Could we automate how the package-specific versions get populated?

I think it's impossible to automate _populating_ the versions without implementing special support in conda-smithy.

> Or at least auto-validate it based on some in-repo information?

Yes, it is validated in build.sh, specifically to avoid these accidental incorrect versions (you're right, it is technically SOVERSION). suitesparse uses default cmake behavior, where SOVERSION is set from package VERSION, so these are necessarily the same (it's a lot more complicated and more likely to differ with autotools). That technically _could_ change, so I could update the check to extract the cmake package version directly, if you prefer. I'll look into that.

File: recipe/build-one.bat
Comment (minrk) [reply recipe/build-one.bat:10]: I can do that. I was planning to do that in #100 as soon as this is merged, but I can merge it here if that's better.

File: recipe/build-one.sh
Comment (minrk) [reply recipe/build-one.sh]: I updated this to check the CMAKE_PROJECT_VERSION directly instead of SOVERSION

File: recipe/build-one.bat
Comment (minrk) [reply recipe/build-one.bat]: this now includes the flang19 migration and uses the simpler `"%FC%" NEQ ""` condition

**Pull request merged**
